croquemort/tools.py

```diff
@@ -75,17 +51,66 @@ def extract_filters(querystring_dict):
     return filters, excludes
 
 
-def apply_filters(results, filters, excludes):
-    """Return filtered results."""
-    if filters:
-        if all(results.get(prop) == value
-               for prop, value in filters.items()):
-            return results
+def apply_filters(data, filters, excludes):
+    """Return filtered data."""
+    filters = filters.copy()
+    excludes = excludes.copy()
+    has_domain_filter = 'domain' in filters
+    has_domain_exclude = 'domain' in excludes
+    has_domain = has_domain_filter or has_domain_exclude
+    filtered_domain = (
+        has_domain_filter
+        and urlparse(data['url']).netloc == filters.pop('domain'))
+    excluded_domain = (
+        has_domain_exclude
+        and urlparse(data['url']).netloc == excludes.pop('domain'))
+    kept_domain = ((has_domain_filter and filtered_domain)
+                   or (has_domain_exclude and not excluded_domain))
+    # Only filter the domain and quick return if no other filters.
+    if has_domain and not (filters or excludes):
+        if kept_domain:
+            return data
+        else:
+            return
+
+    has_props = all(data.get(prop) == value
+                    for prop, value in filters.items())
+    has_not_props = all(data.get(prop) != value
+                        and data.get(prop) is not None
+                        for prop, value in excludes.items()
+                        if prop in data)
+    if filters and excludes:
```

Comment:
If only expressing the empty return case, can by simplified into:
```python
if filters and excludes:
    if has_props and has_not_props:
        if has_domain and not kept_domain:
            return
    else:
        return
elif filters:
    if has_props:
        if has_domain and not kept_domain:
            return
    else:
        return
elif excludes:
    if has_not_props:
        if has_domain and not kept_domain:
            return
    else:
        return

return data
```

Then negating some conditions and using `elif` permit to drop one imbrication level:
```python
if filters and excludes:
    if not has_props or not has_not_props:
        return
    elif has_domain and not kept_domain:
        return
elif filters:
    if not has_props:
        return
    elif has_domain and not kept_domain:
        return
elif excludes:
    if not has_not_props:
        return
    elif has_domain and not kept_domain:
        return

return data
```

I don't know if it's clearer.